libass/ass_font.c

```diff
@@ -35,6 +42,197 @@
 #include "ass_utils.h"
 #include "ass_shaper.h"
 
+
+/**
+ * Convert a UCS-4 code unit to a packed uint32_t in given multibyte encoding
+ */
+#if defined(_WIN32)
+static uint32_t convert_unicode_to_mb(FT_Encoding encoding, uint32_t codepoint)
+{
+    uint32_t result = codepoint;
+    uint16_t input_buffer[2];
+    char output_buffer[4];
+    size_t inbuf_size;
+    UINT codepage;
+    BOOL conversion_fail;
+
+    // encode surrogate pair, assuming codepoint > 0 && codepoint <= 10FFFF
+    if (codepoint >= 0x10000) {
+        // surrogate pair required
+        inbuf_size = 2;
+        uint32_t code = (codepoint - 0x10000);
+        input_buffer[0] = 0xD800 | (code >> 10);
+        input_buffer[1] = 0xDC00 | (code & 0x3FF);
+    } else {
+        inbuf_size = 1;
+        input_buffer[0] = codepoint;
+    }
+
+    // map freetype encoding to windows codepage
+    switch (encoding) {
+    case FT_ENCODING_MS_SJIS:
+        codepage = 932;
+        break;
+    case FT_ENCODING_MS_GB2312:
+        codepage = 936;
+        break;
+    case FT_ENCODING_MS_BIG5:
+        codepage = 950;
+        break;
+    case FT_ENCODING_MS_WANSUNG:
+        codepage = 949;
+        break;
+    case FT_ENCODING_MS_JOHAB:
+        codepage = 1361;
+        break;
+    default:
+        return codepoint;
+    }
+
+    // do convert
+    int output_size = WideCharToMultiByte(
+            codepage,
+            0,
+            input_buffer,
+            inbuf_size,
+            output_buffer,
+            4,
+            NULL,
+            &conversion_fail);
```

Comment:
Please don't use a magic `4` but `sizeof(output_buffer)`. Also setting the `WC_NO_BEST_FIT_CHARS`-flag might be good.
Finally, I'd prefer the arguments to be placed on one or two lines, not one line per argument.